### ISSUE TITLE
Figure out persistence/backup for our juptyerhub accounts

### ISSUE THREAD

**Issue by phaustin** [eoas-ubc/eoas_tlef#47]
[ansible with restc](https://janikvonrotz.ch/2020/03/16/backup-docker-volumes-with-ansible-and-restic/)

[ansible docker copy](https://stackoverflow.com/questions/55166026/ansible-playbook-copy-file-to-docker-container)

**Comment by slemonide** [comment]
Would simply coping `/var/lib/docker/volumes/` somewhere safe be enough?

**Comment by phaustin** [comment]
sure, for a first step -- like that docker copy link above
when we start putting students on the 3 raided nodes in our M235 machine room, we'll need something more like that restc approach

**Comment by phaustin** [comment]
Do you know if there's any buffering -- or if stopping the hub will flush all the open volumes to /var/lib/docker/volumes?

**Comment by phaustin** [comment]
(i.e. -- I guess you can't simple restore by untarring volumes.tar?)

**Comment by slemonide** [comment]
I think it should work if containers that use those volumes are stopped.

For jupyter specifically, it simply stores notebooks in text files. So, I would expect it to behave like any text editor. All closed files should be saved. Any open/unsaved files won't be saved (only their old versions that are persisted to disk will). If files are being saved while they are copied,  they will be corrupted. Though, it is possible to "lock" the files from modifications before copying them.

**Comment by slemonide** [comment]
At the very least with "local" storage drivers (on linux), volumes seem to behave just like any other host system directories, so there shouldn't be any additional buffering and flushing.

**Comment by phaustin** [comment]
thanks, doing this now

**Comment by phaustin** [comment]
I'll make sure the students know that from now on they will need to save their work in a folder called "backups" or something, to avoid having to copy the git repo that is being cloned for every user.

**Comment by phaustin** [comment]
https://docker-py.readthedocs.io/en/stable/containers.html#  -- get_archive and put_archive

**Comment by slemonide** [comment]
> I'll make sure the students know that from now on they will need to save their work in a folder called "backups" or something, to avoid having to copy the git repo that is being cloned for every user.

Is the git repo really that large?

**Comment by phaustin** [comment]
right -- not  necessarily the git repo, but modified satellite images and figures from class exercises etc. are going to be an issue as we scale up:

```
4.0K    _data/data_share
4.0K    _data/a301_lib
4.0K    _data/sat_data
16K     _data/.ipynb_checkpoints
8.0K    _data/__pycache__
52K     _data/a301_2020/.git/hooks
8.0K    _data/a301_2020/.git/logs/refs/heads
8.0K    _data/a301_2020/.git/logs/refs/remotes/origin
12K     _data/a301_2020/.git/logs/refs/remotes
24K     _data/a301_2020/.git/logs/refs
32K     _data/a301_2020/.git/logs
8.0K    _data/a301_2020/.git/refs/heads
4.0K    _data/a301_2020/.git/refs/tags
8.0K    _data/a301_2020/.git/refs/remotes/origin
12K     _data/a301_2020/.git/refs/remotes
28K     _data/a301_2020/.git/refs
16M     _data/a301_2020/.git/objects/pack
4.0K    _data/a301_2020/.git/objects/info
16M     _data/a301_2020/.git/objects
8.0K    _data/a301_2020/.git/info
4.0K    _data/a301_2020/.git/branches
16M     _data/a301_2020/.git
4.0K    _data/a301_2020/data_share
8.0K    _data/a301_2020/a301_lib
4.0K    _data/a301_2020/sat_data
36K     _data/a301_2020/notebooks/week3/.ipynb_checkpoints
88K     _data/a301_2020/notebooks/week3
17M     _data/a301_2020/notebooks/figures
24K     _data/a301_2020/notebooks/week1
84K     _data/a301_2020/notebooks/week2/.ipynb_checkpoints
8.0K    _data/a301_2020/notebooks/week2/__pycache__
232K    _data/a301_2020/notebooks/week2
17M     _data/a301_2020/notebooks
32M     _data/a301_2020
32M     _data
```

**Comment by slemonide** [comment]
I guess it's going to be an issue even if modified files are saved in a separate folder. Since people would probably tend to have a copy of notebook/output for every modification they make (since they wouldn't want to lose previous code/data).

**Comment by phaustin** [comment]
right -- for this course in particular hard quotas are going to have to be enforceable

**Comment by slemonide** [comment]
How much space are you planning to give to each student?

**Comment by phaustin** [comment]
the 3 ibm blades that charles and henryk are refitting for our machine room will have 300 Gbytes Raid 1.  If 15 students can run on 8 cores/16 G ram, then they could have most of one of the blades.  But we wouldn't want to back all of that up.  I'll also be showing them how to install python on their laptops in a couple of weeks, so the number could drop to say 7 students relying on our hub.

**Comment by slemonide** [comment]
8 cores/16 G ram seems very little for 15 students. But it depends on what they do.

**Comment by phaustin** [comment]
we always have the option of adding a new container image to jupyterhub.eoas.ubc.ca once we get persistence/backups set up for kubernetes.   We haven't been given a budget, but storage costs are obviously the biggest expense on AWS, so we'd need to monitor disk use very carefully/figure out how to limit them to say 2 Gbytes.

**Comment by phaustin** [comment]
We could also spread them out on the other 2 nodes -- not many other users until term 2.

**Comment by slemonide** [comment]
Which parts are hosted on AWS and which ones are on-premises?

**Comment by slemonide** [comment]
> storage costs are obviously the biggest expense on AWS

Where do they say that?

**Comment by phaustin** [comment]
disk cost -- that's our experience with research hosting -- they use the instance pricing as a lure, and make their money on storage.  Charles is running some cost summaries on Monday for our EOSC 350 jupyterhub.

Currently:  jupyterhub.eoas.ubc.ca:  kubernetes/EKS in the AWS canadian region
                   eoas1-5:  racked in the UDC
                   eoas6-8:  will be racked in EOAS Main 235 once the redundant power supplies and drives are installed

**Comment by slemonide** [comment]
AWS probably isn't the cheapest hosting provider

**Comment by slemonide** [comment]
If network is fast, storage can be mounted remotely. And the actual data can be stored on-premises.

I think hard drives are very cheap now. Can get 10 TB hard drive for less than 400 CAD, with price of less than 4 CAD per 100 GB.

**Comment by slemonide** [comment]
Also, amazon S3 is probably cheaper for storage than EKS.

**Comment by phaustin** [comment]
that's true, but they won the competition for UBC's business, so they're 
what we've got, unless we want to collect student fees for computing 
(which is illegal without UBC Senate approval it turns out)

**Comment by phaustin** [comment]
the problem with building out more of our own nodes is that our machine room  doesn't meet UBC data security standards (also power/air conditioning is limited) so we are really forced into the cloud.  But even $30/term would be a bargain for students, since they don't have to buy textbooks for these courses.

**Comment by slemonide** [comment]
Then we can stilly try using Amason S3. It seems cheaper.

Amazon S3: $0.023 per GB per month
EKS storage:  $0.004445 per GB per hour = $3.2004 per GB per month

So, there's two order of magnitude difference in price.

I think S3 can be simply bind mounted as a volume.

**Comment by slemonide** [comment]
https://jlelse.blog/dev/restic-docker/

**Comment by slemonide** [comment]
We don't really have to use ansible to configure backups. In "ansible with restc" link, they just use ansible to set up every docker container.

But it's perfectly fine to just run restic in a normal docker container with docker-compose.

On the other hand, it might be useful to be able to setup the whole infrastructure of containers on the nodes using ansible.

But docker-compose seems simpler.

**Comment by slemonide** [comment]
Restic can backup to a standalone restic server, or it can also backup to a variety of other destinations in the cloud, including Amazon S3.

It can also backup using simple sftp, so nothing would need to be installed on the remote server.

There is a long list of supported storage providers here:
* https://restic.readthedocs.io/en/stable/030_preparing_a_new_repo.html#
* https://rclone.org/

**Comment by slemonide** [comment]
Can even configure it to stop notebooks before doing backups to prevent corrupted data: https://github.com/djmaze/resticker

**Comment by phaustin** [comment]
nice -- why don't we start with a local backup that uses resticker, then we can see what charles things about a rest api
